### PR TITLE
Text Editors - Add animated labels

### PR DESCRIPTION
**Pull request by groshenkovamarina**
https://github.com/DevExpress/DevExtreme/pull/19224

### CODE REVIEW

**Review by alexander-kotov-dx — COMMENTED** [review]
no top border above the editor buttons
no-top-border

**Review by alexander-kotov-dx — CHANGES_REQUESTED** [review]
not enougth place for label for selectBox in Generic theme
selectbox-label

And also DropDownBoxes can open the popup to the top in some cases. It should cover part of label?
label-under-dropdown

File: js/ui/text_box/ui.text_editor.base.js [hunk 1]
```diff
@@ -201,9 +210,11 @@ const TextEditorBase = Editor.inherit({
         this.callBase();
 
         this._renderValue();
+        this._renderLabel();
     },
 
     _render: function() {
+        this._renderLabel();
```
Comment (alexander-kotov-dx) [inline js/ui/text_box/ui.text_editor.base.js]: Is it necessary?

File: js/ui/text_box/ui.text_editor.base.js [hunk 2]
```diff
@@ -201,9 +210,11 @@ const TextEditorBase = Editor.inherit({
         this.callBase();
 
         this._renderValue();
+        this._renderLabel();
```
Comment (alexander-kotov-dx) [inline js/ui/text_box/ui.text_editor.base.js:213]: initMarkup changes should be tested in serverSide tests (markup.tests.js)

File: testing/tests/DevExpress.ui.widgets.editors/textEditorParts/common.tests.js [hunk 3]
```diff
@@ -465,6 +468,70 @@ QUnit.module('general', {}, () => {
...
+        const $container = $('#texteditor');
+        const textEditor = $container
+            .dxTextEditor()
+            .dxTextEditor('instance');
+
+        assert.equal($container.find('.' + TEXTEDITOR_LABEL_CLASS).length, 0, 'a label does not render if the option is not specified');
+        assert.notOk($container.hasClass(TEXTEDITOR_WITH_LABEL_CLASS), 'container does not have static class');
+        assert.notOk($container.hasClass(TEXTEDITOR_WITH_FLOATING_LABEL_CLASS), 'container does not have floating class');
+
+        textEditor.option('label', 'label');
```
Comment (alexander-kotov-dx) [inline testing/tests/DevExpress.ui.widgets.editors/textEditorParts/common.tests.js:482]: Please use the separate test for every test scenario

**Review by alexander-kotov-dx — COMMENTED** [review]
We should describe new options in d.ts

**Review by alexander-kotov-dx — COMMENTED** [review]
Colorbox without value has wrong label position
colorbox-without-value

**Review by alexander-kotov-dx — COMMENTED** [review]
label position looks strange in this scenario
vertical-position

File: js/ui/text_box/ui.text_editor.base.js [hunk 4]
```diff
@@ -432,6 +443,27 @@ const TextEditorBase = Editor.inherit({
...
     },
 
+    _renderLabel: function(options) {
+        if(this._$label) {
+            this._$label.remove();
+            this._$label = null;
+        }
+
+        this.$element().removeClass(TEXTEDITOR_WITH_LABEL_CLASS);
+        this.$element().removeClass(TEXTEDITOR_WITH_FLOATING_LABEL_CLASS);
```
Comment (alexander-kotov-dx) [inline js/ui/text_box/ui.text_editor.base.js]: this.$element()
    .removeClass(TEXTEDITOR_WITH_LABEL_CLASS)
    .removeClass(TEXTEDITOR_WITH_FLOATING_LABEL_CLASS);

File: js/ui/text_box/ui.text_editor.base.js [hunk 5]
```diff
@@ -432,6 +443,27 @@ const TextEditorBase = Editor.inherit({
...
+
+        if(!this.option('label') || this.option('labelMode') === 'hidden') return;
+
+        this.$element().addClass(this.option('labelMode') === 'floating' ? TEXTEDITOR_WITH_FLOATING_LABEL_CLASS : TEXTEDITOR_WITH_LABEL_CLASS);
+
+        const labelText = this.option('label');
+        const $label = this._$label = $('<div>')
```
Comment (alexander-kotov-dx) [inline js/ui/text_box/ui.text_editor.base.js:474]:
```
const $labelText = $('<span>').text(this.option('label'));
this._$label = $('<div>')
    .addClass(TEXTEDITOR_LABEL_CLASS)
    .append($labelText)

this._$label.appendTo(this._$textEditorInputContainer);
```
?

**Comment by alexander-kotov-dx** [comment]
I think the label should not change position if the editor has readOnly option

**Comment by alexander-kotov-dx** [comment]
Wrong label position for editors with custom height
custom-height

**Review by babich-a — CHANGES_REQUESTED** [review]
There are no changes in ts declarations. 'label' and 'labelMode' options won't work in modular approaches

File: js/ui/text_box/ui.text_editor.base.js [hunk 6]
```diff
@@ -438,6 +448,45 @@ const TextEditorBase = Editor.inherit({
...
+            .addClass(TEXTEDITOR_LABEL_CLASS)
+            .html('<span>' + labelText + '</span>');
+
+        const $textEditorInputContainer = $container.find('.' + TEXTEDITOR_INPUT_CONTAINER_CLASS);
+
+        this.NAME === 'dxLookup' ?
+            $label.appendTo(container) :
+            $label.appendTo($textEditorInputContainer);
```
Comment (DokaRus) [inline js/ui/text_box/ui.text_editor.base.js:482]: Why not extract the label as a separate component to which to pass the required parameters?

Using a descendant component in a base component doesn't look good

**Pull request closed without merging**